### DISCUSSION TITLE
Support for configuring logic around worker exit to support Prometheus client multiprocessing

### DISCUSSION THREAD

**Discussion by Dominick-Peluso-Bose** [emmett-framework/granian#43] [Ideas]
If you are using granian with multiple workers, I think to make use of Prometheus client we would need a way to hook into the lifecycle of when a worker exits to [clean up Prometheus client multiprocessing](https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn).

In gunicorn, you add something like this to your gunicorn configuration file:

```py
from prometheus_client import multiprocess

def child_exit(server, worker):
    multiprocess.mark_process_dead(worker.pid)
```

It would be great if granian added a way to hook into events like that. Another helpful one would be when the server starts.

**Comment by gi0baro** [comment]
That sounds smth you can already do overriding the spawning method thanks to the `spawn_target` parameter of `Granian.serve`, like:

```python
import os

from granian import Granian
from prometheus_client import multiprocess

def custom_spawn(*args, **kwargs):
    Granian._spawn_wsgi_worker(*args, **kwargs)
    multiprocess.work_process_dead(os.getpid())

if __name__ == "__main__":
    Granian("mymodule:app", interface="wsgi").serve(spawn_target=custom_spawn)
```

of course you can switch wsgi with asgi and rsgi interfaces.

But my question would be: since granian doesn't re-spawn processes (at least right now) is this needed? I mean, if the worker is shutting down, it also implicitly means the entire server/app is shutting down, so you still won't produce any metrics in prometheus..

**Comment by thomas-touhey** [comment]
Hi @gi0baro, from what I get Granian does respawn workers now, and I've been trying to accomplish a similar goal as the OP (call `mark_process_dead` when a worker *process* specifically shuts down, not a worker thread).

I've tried overriding at different levels, but unfortunately haven't managed to get it to work, including using the `spawn_target` kwarg in `Server.serve()`, which is too far from the actual worker spawning to be useful.

The easiest change on the user side would be to have a `on_worker_process_exit` hook, only used by `MPServer` (and *not* by `MTServer`), with the PID of the worker process that has exited.

Another set of changes that would make it possible but harder would be:

* Make the worker class a class var on the server, e.g. `WORKER_CLASS: ClassVar[type[AbstractWorker]] = WorkerProcess`, then use it in `_spawn_worker`. This way, I could make a child class from that class, and override one of its methods.
* Move the logic in [this block](https://github.com/emmett-framework/granian/blob/master/granian/server/common.py#L360-L369) to e.g. `AbstractWorker.terminate`. This way, I could override the `terminate` method in a way that I know, if the parent method ends, the worker has been terminated, and I can mark the process as dead.

You can see more details about why this is required and the recommandations with gunicorn [here](https://prometheus.github.io/client_python/multiprocess/).

Thanks :)

**Reply by gi0baro** [reply]
> I've tried overriding at different levels, but unfortunately haven't managed to get it to work, including using the spawn_target kwarg in Server.serve(), which is too far from the actual worker spawning to be useful.

It's not clear to me why you can't use a custom spawn like in the example I provided. wdym by *too far from the actual worker spawning*?

**Reply by thomas-touhey** [reply]
By trying this example (named `try_granian.py`):

```python
from functools import partial
import os

from fastapi import FastAPI
from granian import Granian
from granian.server import MPServer


app = FastAPI()


if __name__ == "__main__":
    custom_spawn = None
    if issubclass(Granian, MPServer):
        def custom_spawn(*args, **kwargs):
            Granian._spawn_asgi_worker(*args, **kwargs)
            print(f"!!! TERMINATING PROCESS {os.getpid()}")

    server = Granian(
        "try_granian:app",
        address="127.0.0.1",
        port=8080,
        interface="asgi",
        reload=True,
    )

    server.serve(spawn_target=custom_spawn)
```

I get this output:

```
[INFO] Starting granian (main PID: 2037513)
[INFO] Listening at: http://127.0.0.1:8080
[INFO] Spawning worker-1 with PID: 2037523
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=9, pipe_handle=11)
                                                  ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ttouhey/.asdf/installs/python/3.14.2/lib/python3.14/multiprocessing/spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/home/ttouhey/.asdf/installs/python/3.14.2/lib/python3.14/multiprocessing/spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
AttributeError: module '__mp_main__' has no attribute 'custom_spawn'
[ERROR] Unexpected exit from worker-1
[INFO] Shutting down granian
[INFO] Stopped worker-1
[INFO] Granian shutdown completed, see ya!
```

**Reply by gi0baro** [reply]
@thomas-touhey use a dedicated `main.py`/`__main__.py` file with relevant stuff defined in the globals:

```python
from granian import Granian


def custom_spawn(*args, **kwargs):
    Granian._spawn_asgi_worker(*args, **kwargs)
    print(f"!!! TERMINATING PROCESS {os.getpid()}")

server = Granian(
    "try_granian:app",
    address="127.0.0.1",
    port=8080,
    interface="asgi",
    reload=True,
)

if __name__ == "__main__":
    server.serve(spawn_target=custom_spawn)
```

**Reply by thomas-touhey** [reply]
Indeed, trying this seems to work:

```
from functools import partial
import os

from granian import Granian
from granian.server import MPServer


custom_spawn = None
if issubclass(Granian, MPServer):
    def custom_spawn(*args, **kwargs):
        Granian._spawn_asgi_worker(*args, **kwargs)
        print(f"!!! TERMINATING PROCESS {os.getpid()}")

if __name__ == "__main__":
    server = Granian(
        "try_granian_asgi:app",
        address="127.0.0.1",
        port=8080,
        interface="asgi",
        reload=True,
    )

    server.serve(spawn_target=custom_spawn)
```

Thanks :)
